src/background_workers.c

```diff
@@ -245,130 +308,126 @@ void *RedisAI_Run_ThreadMain(void *arg) {
-          // occurred in any worker, we just move on
-          if (dag_device_complete || dagError) {
-            RedisModule_Free(evicted_items[i]);
-          }
+      // If there was no progress on the DAG, meaning the inputs to the
+      // current operation were not ready (they depend on other workers on other
+      // queues completing their job), then we put the entry back on the queue
+      if (do_retry == 1) {
+        RedisAI_RunInfo *evicted_rinfo = (RedisAI_RunInfo *)(evicted_items[0]->value);
+
+        if (queueLength(run_queue_info->run_queue) > 0) {
+          // Pop the next item in the queue
+          queueItem *next_item = queuePop(run_queue_info->run_queue);
+          RedisAI_RunInfo *next_rinfo = (RedisAI_RunInfo *)next_item->value;
+          // Push the DAG to the front of the queue, and then the item we just
+          // popped in front of it, so that it becomes the first item in the queue.
+          // The rational is, since the DAG needs to wait for other workers, we are
+          // giving way to the next item and we'll get back to the DAG when that is done
+          queuePushFront(run_queue_info->run_queue, evicted_rinfo);
+          queuePushFront(run_queue_info->run_queue, next_rinfo);
         }
-        // If we're done processing a modelrun or scriptrun command
+        // If there's nothing else in the queue
         else {
-          // If the client is still connected
-          if (evicted_rinfo->client) {
-            // Unblock the client. Note that we do this for every evicted info in a batch,
-            // which will be associated with a different client. Here we unblock every
-            // client appropriately
-            RedisModule_UnblockClient(evicted_rinfo->client, evicted_rinfo);
-          }
+          // We push the DAG back at the front
+          queuePushFront(run_queue_info->run_queue, evicted_rinfo);
+          // Sleep for a millisecond since there's nothing else on the queue that has a
+          // chance to run, we give other workers a chance to produce the inputs needed
+          // for this DAG step
+          usleep(1000);
+        }
+      }
+
+      // If the op was ran successfully, without any error, then put the entry back
+      // on the queue (since the entry is a DAG, there may be more to do)
+      if (do_run == 1 && run_error == 0) {
+        // TODO: possible optimization could be detect if dag is over and we don't push at front
+        // Here we iterate backwards to keep the first evicted on top
+        // A side effect of this is that we are potentially changing priority in the queue
+        // We could solve this using a priority queue, TODO for later
+        for (long long i = array_len(evicted_items) - 1; i >= 0; i--) {
+          // Get the current evicted run info
+          RedisAI_RunInfo *evicted_rinfo = (RedisAI_RunInfo *)(evicted_items[i]->value);
+          queuePushFront(run_queue_info->run_queue, evicted_rinfo);
```

Comment:
A suggestion:
Have an atomic counter for completed dag ops in the dag info. once an op is completed decrease this counter.
Have a map (dictionary?) of <device_str, arr<DagOp>> in the dag info, so you can easily check if there is any dag op for this device. Moreover, as far as I understand, the current op lookup on the worker thread of the device is done when the queue lock is owned by the worker, so you can change the value of the mapping without worries.
Here you can call a function like `Dag_IsComplete` or similar on all the evicted items and also `Dag_DeviceComplete`, and re-enter the evicted items according to their results.
This also can save you a lot of lookup time for the next DagOp of the device, as well as you will not have to re-insert done dags, or device done dag ops.